### PR TITLE
refactor(workflow-designer-ui): extract and tokenize node handle/label components

### PR DESCRIPTION
**Pull request by shige**
## Summary
- Extract reusable `NodeHandleDot` and `NodeInputLabel` components from editor node implementations
- Tokenize node colors in the design system for consistency
- Apply unified components to both old and new workflow editors
- Fix input connection state logic to correctly derive `isConnected` from `connectedInputIds`

### Component extraction
- **NodeHandleDot**: Handles node input/output connection dots with proper state-based styling
- **NodeInputLabel**: Manages input label positioning and text color based on connection state

### Design system
- Migrated node color primitives to `tokens.css` using semantic HSL-based tokens
- Updated `globals.css` to reference new tokens instead of hardcoded values
- Ensures visual consistency across all node types (text, generation, action, trigger, etc.)

### Bug fixes
- Compute `isConnected` from `connectedInputIds` instead of hardcoded `false`
- Fix output handle stroke to only show when connected
- Correct input handle fill to only appear when connected

### Code quality
- Remove duplicate inline styles by centralizing in shared components
- Apply consistent connection state logic across old and new editors
- Format code with Biome
- Address copilot and CodeRabbit feedback

## Testing
https://github.com/giselles-ai/giselle/pull/2036#issuecomment-3434880959

## Other Information

___

### **PR Type**
Enhancement, Bug fix

___

### **Description**
- Extract reusable `NodeHandleDot` and `NodeInputLabel` components for consistent node UI

- Tokenize node colors in design system for maintainability and consistency

- Fix input connection state logic to derive `isConnected` from `connectedInputIds`

- Apply unified components to both old and new workflow editors

- Update test fixtures to use proper ID generators instead of hardcoded strings

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Node Color Tokens<br/>tokens.css"] -- "referenced by" --> B["globals.css<br/>CSS Variables"]
  B -- "used by" --> C["Old Editor<br/>node.tsx"]
  B -- "used by" --> D["New Editor<br/>node.tsx"]
  E["NodeHandleDot<br/>Component"] -- "renders" --> C
  E -- "renders" --> D
  F["NodeInputLabel<br/>Component"] -- "renders" --> C
  F -- "renders" --> D
  G["Test Fixtures<br/>build-levels.test.ts"] -- "uses" --> H["ID Generators<br/>NodeId, OutputId, etc"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>node-handle-dot.tsx</strong><dd><code>New reusable NodeHandleDot component for connection dots</code>&nbsp; </dd></summary>
<hr>

internal-packages/workflow-designer-ui/src/ui/node/node-handle-dot.tsx

<ul><li>New component that encapsulates node input/output connection dot <br>rendering<br> <li> Handles state-based styling with connected/disconnected states<br> <li> Supports multiple content types with appropriate color tokens<br> <li> Includes fallback handling for unknown content types with dev warnings</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-dfb3830c5020286303c491a7435b7a7303930f71a880c61d8c3f416834544a81">+103/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>node-input-label.tsx</strong><dd><code>New reusable NodeInputLabel component for input labels</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal-packages/workflow-designer-ui/src/ui/node/node-input-label.tsx

<ul><li>New component that manages input label positioning and styling<br> <li> Applies different styles based on connection state and required status<br> <li> Shows red text for required disconnected inputs<br> <li> Positions label absolutely when disconnected, inline when connected</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-0810a4f7d1ea62606c93e7ef43249a0241ec8611271d3215714957b36b271329">+27/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>node.tsx</strong><dd><code>Refactor old editor to use extracted node components</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal-packages/workflow-designer-ui/src/editor/node/node.tsx

<ul><li>Replace inline Handle elements with <code>NodeHandleDot</code> component for inputs<br> <li> Replace inline label divs with <code>NodeInputLabel</code> component<br> <li> Add <code>getInputHandleContentType()</code> function to safely map node content <br>types<br> <li> Compute <code>isConnected</code> state from <code>connectedInputIds</code> for both action and <br>non-action inputs<br> <li> Fix GitHub node icon text color from black to inverse</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-0f40b8e983ec8fd4a8a77a5229e17343f84d9bbc676847607b0b7fd895e6210f">+88/-44</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>node.tsx</strong><dd><code>Refactor new editor to use extracted node components</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal-packages/workflow-designer-ui/src/new-editor/components/node/node.tsx

<ul><li>Add <code>connectedInputIds</code> to store selector and component props<br> <li> Add <code>inputHandleContentType</code> computation with memoization<br> <li> Replace inline Handle elements with <code>NodeHandleDot</code> component<br> <li> Replace inline label divs with <code>NodeInputLabel</code> component<br> <li> Update output handle styling to only show border when connected</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-83fb8cf26f413024d1112a94404307f546ffb76041277c3bff08dc1572a70c9d">+107/-61</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>tokens.css</strong><dd><code>Add comprehensive node color tokens to design system</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal-packages/ui/styles/tokens.css

<ul><li>Add 26 new node color tokens in HSL format for all node types<br> <li> Includes primary and secondary colors for each node type<br> <li> Covers text, file, web-page, generation, trigger, action, github, <br>web-search, image-generation, audio-generation, video-generation, <br>query, and vector-store nodes<br> <li> Enables centralized color management and consistency</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-89ac25d4438a7bd19dc58d210fcf310e9d621fea28f95b544c540906bebb5311">+28/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>globals.css</strong><dd><code>Update globals.css to reference tokenized node colors</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/studio.giselles.ai/app/globals.css

<ul><li>Replace hardcoded HSL color values with CSS variable references<br> <li> Update all node color definitions to use new tokens from <code>tokens.css</code><br> <li> Maintains visual consistency while improving maintainability<br> <li> Covers 13 different node type color pairs</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-03a164f1b9b1c0168d663c7c6f996637ee967c1cd2bbda6331d37178f9f72a87">+26/-26</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>next-env.d.ts</strong><dd><code>Add Next.js routes type reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/playground/next-env.d.ts

<ul><li>Add TypeScript reference to Next.js generated routes type definitions<br> <li> Enables type-safe route handling in the playground application</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-37b289bef52fc31caba34830ca346a46134b37520ce40e328322a6af26288223">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>build-levels.test.ts</strong><dd><code>Refactor test fixtures to use ID generators</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/giselle/src/engine/utils/build-levels.test.ts

<ul><li>Replace hardcoded string IDs with proper ID generator calls <br>(<code>NodeId.generate()</code>, <code>OutputId.generate()</code>, etc)<br> <li> Update all test fixtures to use generated IDs instead of manual <br>strings<br> <li> Improve test maintainability and alignment with production code <br>patterns<br> <li> Covers all test cases including parallel execution, mixed node types, <br>edge cases, and cycle handling</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/2036/files#diff-02bc1ee80ee2c476277dab531c32fff51af4daf106c4716fa61653a3a232b685">+255/-179</a></td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

> [!NOTE]
> Extracts reusable node handle/label components, tokenizes node colors, applies across editors, and updates tests to use generated IDs.
> 
> - **UI (workflow-designer)**:
>   - **Shared components**: Add `NodeHandleDot` and `NodeInputLabel` for connection dots and input labels with connection/required-state styling.
>   - **Editor updates**: Replace inline handles/labels in `src/editor/node/node.tsx` and `src/new-editor/components/node/node.tsx`; compute `isConnected` from `connectedInputIds`; map input handle `contentType` safely; adjust output handle styles to only show border/fill when connected; fix GitHub icon text color to `text-inverse`.
> - **Design Tokens (ui)**:
>   - Add HSL-based node color primitives in `internal-packages/ui/styles/tokens.css` and update `apps/studio.giselles.ai/app/globals.css` to reference them.
> - **Tests**:
>   - Refactor `packages/giselle/src/engine/utils/build-levels.test.ts` to use `NodeId`/`InputId`/`OutputId`/`ConnectionId` generators and modernize fixtures (parallel, mixed types, duplicates, cycles).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 545ced96a956d237f96cf5543af87a7a156548eb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **Style**
  * Reworked node color theming: new theme tokens and variable-based node colors for consistent, theme-aware visuals across node types (text, file, web, generation, trigger, action, GitHub, web-search, image/audio/video, query, vector-store).

* **New Features**
  * Improved node input UI: redesigned handle dots and input labels that reflect connection and required states, and added a node preview mode for alternate rendering.

### CODE REVIEW

File: internal-packages/workflow-designer-ui/src/editor/node/node.tsx [hunk 1]
```diff
@@ -26,11 +26,54 @@ import { useEffect, useMemo, useRef, useTransition } from "react";
...
+		case "file":
+		case "webPage":
+		case "webSearch":
+		case "audioGeneration":
+		case "videoGeneration":
+		case "trigger":
+		case "action":
+		case "query":
+			return contentType;
+		case "vectorStore": {
```
Comment (shige) [inline internal-packages/workflow-designer-ui/src/editor/node/node.tsx:57]: 📝 

This PR sticks to resolving the merge conflict and the immediate type errors, so I haven’t tackled the mismatch between the vector store data model and the UI types.
The underlying fix is to realign what NodeHandleDot expects with the values exposed by NodeLike/VectorStoreContent, and I plan to handle that in a follow-up refactoring PR.

File: internal-packages/workflow-designer-ui/src/new-editor/components/node/node.tsx [hunk 2]
```diff
@@ -178,6 +188,36 @@ function CanvasNode({
...
+			case "file":
+			case "webPage":
+			case "webSearch":
+			case "audioGeneration":
+			case "videoGeneration":
+			case "trigger":
+			case "action":
+			case "query":
+				return contentType;
+			case "vectorStore":
```
Comment (shige) [inline internal-packages/workflow-designer-ui/src/new-editor/components/node/node.tsx:208]: 📝 

This PR sticks to resolving the merge conflict and the immediate type errors, so I haven’t tackled the mismatch between the vector store data model and the UI types.
The underlying fix is to realign what NodeHandleDot expects with the values exposed by NodeLike/VectorStoreContent, and I plan to handle that in a follow-up refactoring PR.

**Comment by shige** [comment]
@kaochannel154 @toyamarinyon 
I cherry-picked only the design-related commits from https://github.com/giselles-ai/giselle/pull/2018 and adjusted it so that the build would pass.
Could you please review it?

File: apps/playground/next-env.d.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 /// <reference types="next" />
 /// <reference types="next/image-types/global" />
+/// <reference path="./.next/types/routes.d.ts" />
```
Comment (toyamarinyon) [inline apps/playground/next-env.d.ts]: I don't think this change is necessary for what I want to do with this pull request, so I would like to delete it if possible.

Comment (shige) [reply apps/playground/next-env.d.ts]: Fixed! 545ced96a956d237f96cf5543af87a7a156548eb

**Comment by shige** [comment]
Thank you! 🚀
